src/transport.cpp

```diff
@@ -0,0 +1,138 @@
+#ifdef TRACE_JSON
+                std::ostringstream str;
+                str << prefix << " #" << id << "# " << name;
+
+                if (request_id > 0 && request_id < std::numeric_limits<message_id>::max()) {
+                    str << " #" << request_id << "#";
+                }
+
+                str << " " << json;
+
+                if (!blob.empty()) {
+                    str << " <raw (" << blob.size() << " bytes)>";
+                }
+
+                log::logf("%s\n\n", str.str().c_str());
+#endif
+            }
+
+            void disconnect() {
+                if (connected.exchange(false)) {
+                    disconnected();
+                }
+            }
+
+            void receive_worker() {
+                for (;;) {
+                    boost::endian::little_uint32_buf_t msg_size;
+                    if (fread(&msg_size, sizeof msg_size, 1, input) != 1) {
+                        break;
+                    }
+
+                    std::string payload(msg_size.value(), '\0');
+                    if (!payload.empty()) {
+                        if (fread(&payload[0], payload.size(), 1, input) != 1) {
+                            break;
+                        }
+                    }
+
+                    auto msg = message::parse(payload);
+                    log_message("==>", msg.id(), msg.request_id(), msg.name(), msg.json_text(), msg.blob());
+                    message_received(msg);
+                }
+
+                disconnect();
+            }
+        }
+
+        boost::signals2::signal<void(const protocol::message&)> message_received;
+
+        boost::signals2::signal<void()> disconnected;
+
+        void initialize() {
+            assert(!input && !output);
+
+            setmode(fileno(stdin), _O_BINARY);
+            setmode(fileno(stdout), _O_BINARY);
+
+            // Duplicate and stash away handles for original stdin & stdout.
+            input = fdopen(dup(fileno(stdin)), "rb");
+            setvbuf(input, NULL, _IONBF, 0);
+            output = fdopen(dup(fileno(stdout)), "wb");
+            setvbuf(output, NULL, _IONBF, 0);
+
+            // Redirect stdin and stdout to the null device, so that any code trying to write directly
+            // to them (instead of via R_WriteConsole) will not interfere with the protocol.
+            freopen(devNull, "rb", stdin);
+            freopen(devNull, "wb", stdout);
+
+            connected = true;
+            std::thread(receive_worker).detach();
+        }
+
+        void send_message(const message& msg) {
+            assert(output);
+
+            log_message("<==", msg.id(), msg.request_id(), msg.name(), msg.json_text(), msg.blob());
+
+            auto& payload = msg.payload();
+            boost::endian::little_uint32_buf_t msg_size(static_cast<uint32_t>(payload.size()));
+
+            std::lock_guard<std::mutex> lock(output_lock);
+            if (fwrite(&msg_size, sizeof msg_size, 1, output) == 1) {
```

Comment:
`fwrite` and `fread` return the number of elements read or written. In this case we're writing 1 element (3rd argument here) of size `sizeof msg_size` (2nd argument); so if write succeeded, the return value will be 1, and if it failed or partially succeeded, it'll be 0.